### DISCUSSION TITLE
Thinking more about the agent registry and tools

### DISCUSSION THREAD

**Discussion by fjuma** [wise-agents/wise-agents#178] [Ideas]
The agent registry is meant to contain information about all of the agents that are available (e.g., names and formal description). The idea is for some sort of coordination agent to be able to make use of the info from the registry along with an LLM to break down a goal into sub-goals and to determine the next agent to invoke.

Let's say that we have agents A, B, C, D, E. 

Let's think of B and C as tools that only agent A is expected to invoke. 

Agents D and E won't need to send messages to B and C directly, but could send messages to agent A (which could in turn invoke B and C).

Should B and C be included in the registry even though the coordination agent should not select B or C as the next agent to handle a sub-goal?

One option could be to indicate in the formal description for B and C that these are tools that A can make use of to achieve its goal but should not be invoked directly.

Another option could be to not include B or C in the registry at all. The description of A should indicate what it is capable of doing. The tools it makes use of to accomplish its goal might not be relevant for the registry. I am leaning towards this option.

Any thoughts?

**Comment by maeste** [comment]
In general I agree with your preference. IOW, the registry should contain only minimal reusable blocks. So if an agent A need to use other tools to complete his goal, they don't need to be exposed to the coordinator.
That said, if those tools are agents (and so distributed) there are more considerations: 
1. staying in your example what if agent B could be useful both to complete the goal of A and the goal of D? Agent B doesn't make sense to be exposed to the coordinator, but it may be exposed to the registry.
2. Another point is I see the registry is (also, maybe mainly) the way agents make them available for other agents in a pure distributed architecture. So how do you suggest to declare this A-B-C composition when A-B-C are on different pods/images/network?
3. Returning to what I said in my first sentences, if B and C are used only by A and not distributed, do we really need B and C to be agents? Or are they "just" tools?

just my 2C

**Reply by fjuma** [reply]
> In general I agree with your preference. IOW, the registry should contain only minimal reusable blocks. So if an agent A need to use other tools to complete his goal, they don't need to be exposed to the coordinator. That said, if those tools are agents (and so distributed) there are more considerations:
> 
> 1. staying in your example what if agent B could be useful both to complete the goal of A and the goal of D? Agent B doesn't make sense to be exposed to the coordinator, but it may be exposed to the registry.

Ok yeah I see your point. In general, there will actually also be other agents that the coordinator shouldn't consider for sub-task assignment (e.g., it shouldn't consider itself or the reasoning agent or the final answer agent to get the result for a sub-task) so we can probably use some sort of boolean value or something to distinguish these types of agents from other agents since they should all be part of the registry. 

> 2. Another point is I see the registry is (also, maybe mainly) the way agents make them available for other agents in a pure distributed architecture. So how do you suggest to declare this A-B-C composition when A-B-C are on different pods/images/network?

Good catch! That's right, these would actually need to be part of the registry.

> 3. Returning to what I said in my first sentences, if B and C are used only by A and not distributed, do we really need B and C to be agents? Or are they "just" tools?

Yeah, for this use case, it probably makes more sense for B and C to be just functions/tools as opposed to agents.

> 
> just my 2C
